File: src/main/java/com/github/asufana/presentationjava/fields/Fields.java
package com.github.asufana.presentationjava.fields;

import static java.util.Optional.*;
import static java.util.stream.Collectors.*;

import java.util.*;
import java.util.function.*;
import java.util.stream.*;

import com.github.asufana.presentationjava.utils.StringBuilder;
import org.apache.commons.lang3.StringUtils;

/** フィールドリスト */
public class Fields extends AbstractCollection<Fields, Field> {
    
    public Fields(final List<Field> list) {
        super(list);
    }
    
    /** 詳細画面向けHTML文字列生成 */
    public String toDescTableHtml() {
        final String layout = descTableDefaultColumnsLayout();
        if (StringUtils.isEmpty(layout)) {
            throw new RuntimeException();
        }
        final String html = String.format("<!-- columnsLayout:\n%s\n-->\n%s\n",
                                          layout,
                                          toDescTableHtml(layout));
        return html;
    }
    
    /** 詳細画面向けデフォルトカラムレイアウト */
    public String descTableDefaultColumnsLayout() {
        final String layout = list.stream()
                                  .map(field -> field.instanceFullName())
                                  .collect(joining("\n"));
        return layout;
    }
    
    /** 詳細画面向けHTML文字列生成 */
    public String toDescTableHtml(final String columnsLayout) {
        //ボディ部
        final String body = toDescBodyString(columnsLayout);
        //テーブルHTML
        final StringBuilder sb = new StringBuilder();
        sb.addLine("<table class=\"table table-bordered table-condensed\">")
          .addLine("<tbody>")
          .add(body)
          .addLine("</tbody>")
          .addLine("</table>");
        return sb.toString();
    }
    
    /** 詳細画面ボディ */
    protected String toDescBodyString(final String columnsLayout) {
        
        //インスタンス名からHTMLに変換する
        //@param instanceNames: ex. field1,field2,field3
        final Function<String, String> instanceNameToHtmlFunction = instanceNames -> {
            final String innerHtml = Stream.of(instanceNames.split(","))
                                           .map(instanceName -> findField(instanceName).orElse(LesserField.create(instanceName)))
                                           .map(field -> field.toHeaderHtml() + field.toBodyHtml())
                                           .collect(joining("\n"));
            return innerHtml;
        };
        
        //レイアウト文字列からHTMLに変換する
        //@param columnsLayout: ex. field1,field2\n field3,field4,field5
        final String html = Stream.of(columnsLayout.split("\n"))
                                  .map(instanceNameToHtmlFunction)
                                  .collect(joining("\n</tr>\n<tr>\n"));
        return String.format("<tr>\n%s\n</tr>\n", html);
    }
    
    /** 一覧画面向けHTML文字列生成 */
    public String toListTableHtml() {
        //ヘッダ部
        final String header = toListHeaderString();
        //ボディ部
        final String body = toListBodyString();
        
        final StringBuilder sb = new StringBuilder();
        sb.addLine("<table class=\"table table-striped table-bordered table-condensed\">")
          .addLine("<thead>")
          .add(header)
          .addLine("</thead>")
          .addLine("<tbody>")
          .add(body)
          .addLine("</thead>")
          .addLine("</tbody>")
          .addLine("</table>");
        return sb.toString();
    }
    
    /** 一覧表示ヘッダ */
    protected String toListHeaderString() {
        final String str = list.stream().map(Field::toHeaderHtml).collect(joining());
        return str;
    }
    
    /** 一覧表示ボディ */
    protected String toListBodyString() {
        final String str = list.stream().map(Field::toBodyHtml).collect(joining());
        return str;
    }
    
    /** インスタンス名からフィールドの取得 */
    public Optional<Field> findField(final String instanceNames) {
        final String s1 = toClean(instanceNames);
        final List<Field> matchedFields = list.stream().filter(field -> {
            final String s2 = toClean(field.instanceFullName());
            //前方一致で抽出
            return s2.startsWith(s1);
        }).collect(toList());
        
        //前方一致で抽出した結果、フィールドが一意に特定されればそれを返却する
        //employee.name.value に対して、
        //employee.name.value でも employee.name でもOKとする
        return matchedFields.size() == 1
                ? of(matchedFields.get(0))
                : Optional.empty();
    }
    
    /** ユーザ入力インスタンス名の正規化 */
    protected String toClean(final String instanceNames) {
        return instanceNames.replaceAll("\\s", "").toLowerCase();
    }
    
    @Override
    public String toString() {
        return list.stream().map(Field::toString).collect(joining("\n"));
    }
}

File: src/main/java/com/github/asufana/presentation/fields/FIelds.kt
package com.github.asufana.presentation.fields


data class Fields(val fields: List<Field>) {

    /** 詳細画面向けHTML文字列生成 */
    fun toDescTableHtml(): String {
        val layout = descTableDefaultColumnLayout()
        if (layout.isEmpty()) throw RuntimeException()

        val tableHtml = toDescTableHtml(layout)
        val html = """
<!-- columnsLayout -->
<!--
$layout
-->
$tableHtml
"""
        return html
    }

    /** 詳細画面向けHTML文字列生成 */
    fun toDescTableHtml(columnsLayout: String): String {
        //ボディ部
        val body = toDescBodyString(columnsLayout)
        //テーブルHTML
        val html = """
<table class="table table-bordered table-condensed">
$body
</table>
"""
        return html
    }

    /** 詳細画面ボディ */
    protected fun toDescBodyString(columnsLayout: String): String {
        //インデント空白
        val indent = "  "

        //インスタンス名からHTMLに変換するファンクション
        //@params instanceNames: ex) field1,field2,field3
        val instanceNameToHtmlFunction: (String) -> String = {
            instanceNames ->
            instanceNames.split(",")
                    .map { it.trim() }
                    //フィールドが見つからない場合にはNullオブジェクトを渡す
                    .map { instanceName -> findField(instanceName) ?: NullField(instanceName) }
                    .map { field -> field.toHeaderHtml() + field.toBodyHtml() }
                    .joinToString("\n$indent$indent")
        }

        //レイアウト文字列からHTMLに変換する
        //@params columnsLayout: ex) field1,field2\n field3,field4,field5
        val html = columnsLayout.split("\n")
                .map { it.trim() }
                .map { instanceNameToHtmlFunction(it) }
                .map { "$indent$indent$it" }
                .joinToString("\n$indent</tr>\n$indent<tr>\n")

        return "$indent<tr>\n$html\n$indent</tr>"
    }

    /** 詳細画面向けデフォルトレイアウト */
    fun descTableDefaultColumnLayout(): String {
        val layout = fields.map { it.instanceFullName }.joinToString("\n")
        return layout
    }

    /** インスタンス名からフィールドの取得 */
    protected fun findField(instanceName: String): Field? {
        val s1 = toClean(instanceName)
        val matchedFields = fields
                .filter {
                    //完全一致ではなく前方一致で判定する
                    //employee.name.value に対して、
                    //employee.name.value でも employee.name でもOKとする
                    it.instanceFullName != null &&
                            toClean(it.instanceFullName).startsWith(s1)
                }

        //前方一致で抽出した結果、フィールドが一意に特定されればそれを返却する
        return if (matchedFields.size == 1) matchedFields[0] else null
    }

    /** ユーザ入力インスタンス名の正規化 */
    protected fun toClean(instanceName: String): String {
        return instanceName.replace("\\s", "").toLowerCase();
    }

    fun size(): Int {
        return fields.size
    }

    override fun toString(): String {
        return fields.map { it.toString() }.joinToString("\n")
    }
}



